### PR TITLE
rgw: switch AES_256_{CBC, ECB_encrypt} from NSS to OpenSSL

### PR DESCRIPTION
**Pull request by rzarzynski**
After this patchset gets in, the last NSS user will be the legacy PKI token support of Keystone.

**UPDATE**: PKI/PKIz  tokens have been dropped in 2016 from Keystone. See https://github.com/openstack/keystone/commit/8a66ef635400083fa426c0daf477038967785caf.

### CODE REVIEW

File: src/rgw/rgw_crypt.cc [hunk 1]
```diff
@@ -16,6 +16,11 @@
 #include "include/str_map.h"
 #include "crypto/crypto_accel.h"
 #include "crypto/crypto_plugin.h"
+
+#ifdef USE_OPENSSL
+# include <openssl/evp.h>
+#endif // USE_OPENSSL
+
```
Comment (mattbenjamin) [inline src/rgw/rgw_crypt.cc]: #else/#elif?

File: src/rgw/rgw_crypt.cc [hunk 2]
```diff
@@ -172,6 +172,74 @@ CryptoAccelRef get_crypto_accel(CephContext *cct)
...
+static inline
+bool evp_sym_transform(CephContext* const cct,
+                       const EVP_CIPHER* const type,
+                       unsigned char* const out,
+                       const unsigned char* const in,
+                       const size_t size,
+                       const unsigned char* const iv,
+                       const unsigned char* const key,
+                       const bool encrypt)
+{
```
Comment (mattbenjamin) [inline src/rgw/rgw_crypt.cc]: is line continuation needed here?

Comment (mattbenjamin) [reply src/rgw/rgw_crypt.cc]: oh, it's not new anyway, sorry

File: src/rgw/rgw_crypt.cc (hunk 1 above)
Comment (rzarzynski) [reply src/rgw/rgw_crypt.cc]: Fixed by keeping the `#else` clause we had for NSS.

File: src/rgw/rgw_crypt.cc (hunk 2 above)
Comment (rzarzynski) [reply src/rgw/rgw_crypt.cc]: Do you mean  new line before the bracket? I tried to be consistent with closest neighborhood but I stay open to reformatting entire file.

**Comment by cbodley** [comment]
looks good to me, and passing tests. ping @mdw-at-linuxbox

**Comment by rzarzynski** [comment]
It's worth to mention that https://github.com/ceph/ceph/pull/27834 is built on top of these patches. In contrast to the broader PR, changes here are focused solely on RGW. This can help with reviewing both of them, I hope.

**Comment by rzarzynski** [comment]
Closing as #27834 has been merged!

**Pull request closed without merging**